coderd/jwtutils/jws.go

```diff
@@ -0,0 +1,129 @@
+package jwtutils
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/go-jose/go-jose/v4"
+	"github.com/go-jose/go-jose/v4/jwt"
+	"golang.org/x/xerrors"
+)
+
+const (
+	keyIDHeaderKey = "kid"
+)
+
+// Claims defines the payload for a JWT. Most callers
+// should embed jwt.Claims
+type Claims interface {
+	Validate(jwt.Expected) error
+}
+
+const (
+	signingAlgo = jose.HS512
+)
+
+type SigningKeyProvider interface {
+	SigningKey(ctx context.Context) (id string, key interface{}, err error)
+}
+
+type VerifyKeyProvider interface {
+	VerifyingKey(ctx context.Context, id string) (key interface{}, err error)
+}
+
+// Sign signs a token and returns it as a string.
+func Sign(ctx context.Context, s SigningKeyProvider, claims Claims) (string, error) {
+	id, key, err := s.SigningKey(ctx)
+	if err != nil {
+		return "", xerrors.Errorf("get signing key: %w", err)
+	}
+
+	signer, err := jose.NewSigner(jose.SigningKey{
+		Algorithm: signingAlgo,
+		Key:       key,
+	}, &jose.SignerOptions{
+		ExtraHeaders: map[jose.HeaderKey]interface{}{
+			keyIDHeaderKey: id,
+		},
+	})
+	if err != nil {
+		return "", xerrors.Errorf("new signer: %w", err)
+	}
+
+	payload, err := json.Marshal(claims)
+	if err != nil {
+		return "", xerrors.Errorf("marshal claims: %w", err)
+	}
+
+	signed, err := signer.Sign(payload)
+	if err != nil {
+		return "", xerrors.Errorf("sign payload: %w", err)
+	}
+
+	compact, err := signed.CompactSerialize()
+	if err != nil {
+		return "", xerrors.Errorf("compact serialize: %w", err)
+	}
+
+	return compact, nil
+}
+
+// VerifyOptions are options for verifying a JWT.
+type VerifyOptions struct {
+	RegisteredClaims jwt.Expected
+
+	// The following are only used for JWSs.
```

Comment:
Can you use `VerifyOptions` for anything other than a JWS? The comment reads like these options can be used for something else.